File: android/app/src/main/java/com/rbdarts/feature/league/LeagueService.kt
package com.rbdarts.feature.league

import com.rbdarts.core.data.RBDartsRepository
import com.rbdarts.core.domain.AverageUpdateRule
import com.rbdarts.core.domain.ExtraInningsRule
import com.rbdarts.core.domain.HandicapResult
import com.rbdarts.core.domain.HandicapRoundingRule
import com.rbdarts.core.domain.League
import com.rbdarts.core.domain.ScoringRules
import com.rbdarts.core.domain.TeamStanding
import com.rbdarts.core.domain.TieBonusRule
import java.time.Instant
import java.util.UUID

class LeagueService(private val repository: RBDartsRepository) {
    suspend fun createLeague(
        name: String,
        gamesPerMatch: Int,
        playersPerTeamPerGame: Int,
        handicapPercent: Double
    ): League {
        val now = Instant.now()
        val league = League(
            leagueId = UUID.randomUUID().toString(),
            leagueName = name,
            gamesPerMatch = maxOf(1, gamesPerMatch),
            playersPerTeamPerGame = maxOf(1, playersPerTeamPerGame),
            handicapPercent = handicapPercent.coerceIn(0.0, 100.0),
            handicapRoundingRule = HandicapRoundingRule.Nearest,
            pointsPerGameWin = 1.0,
            matchBonusPointValue = 1.0,
            averageUpdateRule = AverageUpdateRule.AfterMatch,
            extraInningsRule = ExtraInningsRule.Enabled,
            tieBonusRule = TieBonusRule.NoBonus,
            createdAt = now,
            updatedAt = now
        )
        repository.createLeague(league)
        return league
    }

    fun handicapForGame(
        homeTeamId: String,
        awayTeamId: String,
        homeAverageSum: Double,
        awayAverageSum: Double,
        league: League
    ): HandicapResult = ScoringRules.handicap(
        teamAverageSums = mapOf(homeTeamId to homeAverageSum, awayTeamId to awayAverageSum),
        handicapPercent = league.handicapPercent,
        roundingRule = league.handicapRoundingRule
    )

    fun standings(standings: List<TeamStanding>): List<TeamStanding> =
        ScoringRules.rankedStandings(standings)
}

File: ios/RBDarts/Features/League/LeagueService.swift
import Foundation

public struct LeagueMatchPlan: Equatable, Sendable {
    public let league: League
    public let homeTeam: Team
    public let awayTeam: Team
    public let handicap: HandicapResult
}

public struct LeagueService: Sendable {
    private let repository: RBDartsRepository

    public init(repository: RBDartsRepository = InMemoryRBDartsRepository()) {
        self.repository = repository
    }

    public func createLeague(
        name: String,
        gamesPerMatch: Int,
        playersPerTeamPerGame: Int,
        handicapPercent: Double
    ) async throws -> League {
        let now = Date()
        let league = League(
            leagueId: UUID().uuidString,
            leagueName: name,
            gamesPerMatch: max(1, gamesPerMatch),
            playersPerTeamPerGame: max(1, playersPerTeamPerGame),
            handicapPercent: min(max(handicapPercent, 0), 100),
            handicapRoundingRule: .nearest,
            pointsPerGameWin: 1,
            matchBonusPointValue: 1,
            averageUpdateRule: .afterMatch,
            extraInningsRule: .enabled,
            tieBonusRule: .noBonus,
            createdAt: now,
            updatedAt: now
        )
        try await repository.createLeague(league)
        return league
    }

    public func handicapForGame(
        homeTeamId: String,
        awayTeamId: String,
        homeAverageSum: Double,
        awayAverageSum: Double,
        league: League
    ) -> HandicapResult {
        ScoringRules.handicap(
            teamAverageSums: [homeTeamId: homeAverageSum, awayTeamId: awayAverageSum],
            handicapPercent: league.handicapPercent,
            roundingRule: league.handicapRoundingRule
        )
    }

    public func standings(_ standings: [TeamStanding]) -> [TeamStanding] {
        ScoringRules.rankedStandings(standings)
    }
}
